### ISSUE TITLE
Invalid escape

### ISSUE THREAD

**Issue by mathiasrw** [slevithan/regex#21]
Creating this query "native" works

    console.log('Native: ', new RegExp('(?<Q>[abc])[\k<Q>]', 'v'))

Doing the same with regex throws an "invalid escape"

    import { regex, pattern } from 'regex';
    const JZON_REGEX = regex`(?<Q>[abc])[\k<Q>]`;
    console.log(JZON_REGEX);
    

The output says that 

    Invalid regular expression: /(?<Q>[abc])[\k<Q\>]/v: Invalid escape

Notice the `\` in front of the last `>` - I think there is a auto conversion that goes wong here. 
    
<img width="609" alt="image" src="https://github.com/user-attachments/assets/99316941-a859-4e0e-85cc-38657d6db7a3">

**Comment by slevithan** [comment]
`regex` is correct.

In your version using `RegExp`, you forgot to escape your backslash. If you change it to `console.log('Native: ', new RegExp('(?<Q>[abc])[\\k<Q>]', 'v'))`, it will throw the same error.

In other words, the problem is not the character you're pointing to in your screenshot. It is the `\k`, which is invalid within a character class. You're not seeing the problem in the "native" version because `\` is a special character in non-raw strings, and since `\k` has no special meaning in JavaScript strings (unlike e.g. `\\` or `\n`), the `\` is stripped and what you're actually passing to the `RegExp` constructor is just plain `k`, without the slash (which you can see in the output of `console.log`). To avoid this, you either need to escape your backslashes in the string, or use `String.raw`. On the other hand, `regex` is a raw string template, so you don't need to escape your backslashes in patterns you pass to the `regex` template tag. `regex` uses raw strings to improve readability (avoiding the need to escape all backslashes).

**Comment by slevithan** [comment]
As for why `regex` is escaping `>` within the character class, this is complicated, but note that it doesn't change the meaning at all.

- It is optional to escape `<` and `>` within character classes when using flag `v`, so doing so doesn't change their meaning unless they are followed by the same character. `/[<]/v` and `/[>]/v` are fine, but `/[<<]/v` and `/[>>]/v` are both errors since these are "reserved double punctuators".
- It is **invalid** to escape `<` and `>` within character classes when using **flag `u`**, so `regex` doesn't do so when running in environments without native `v`, or when you explicitly disable `v`. And since, by default, `regex` applies `v`'s escaping rules even when using flag `u` (assuming you leave the default value for option `unicodeSetsPlugin`), it is always fine to escape these characters within character classes. When using flag `u`, `regex` will automatically unescape them (which means you can always use `v`'s escaping rules, even in environments without native `v`). Also, even when using flag `u`, `regex` will continue to throw a "reserved double punctuator" error for `<<` and `>>` within character classes.
- The *reason* for the escaping is an implementation detail related to `regex`'s sandboxing of tokens within character classes when using implicit flag `x`, which allows insignificant <kbd>space</kbd> and <kbd>tab</kbd> within character classes (following the rules of Perl/PCRE's `xx` flag). But whitespace still separates tokens and is not simply ignored. This means that `` regex`[>]` `` and `` regex`[> >]` `` are both equivalent to `/[>]/v`. But `` regex`[>>]` `` (without the separating space) correctly returns an error, as does `/[>>]/v`.

All of this in fact shows `regex`'s deep understanding of all the nuances/complexities of regex syntax in all modes and with all flags, which is COMPLICATED. 😖

It's possible that, in future versions, `regex` will avoid escaping `>` with the input you gave, but this will not change the meaning of the regex.

**Comment by slevithan** [comment]
Closing this since there's no bug in the reported input, but I'm happy to continue discussing if helpful, and I created #22 to follow up on `regex`'s eagerness of escaping certain characters.

**Comment by mathiasrw** [comment]
Thank you for looking info this. Hmm. I see. So many things to learn. Escapes are nasty. 

**Are you saying that with the `v` flag I can no longer include a back reference inside a character class ?**

Im happy its not the regex lib that has a but, but on the other hand I felt a bit of excitement when I thought I had found a way to help improve this lib. 


---


I would love to pick your brain on the original problem, if you dont mind. I would like to match quoted strings ignoring the quote when its being escaped itself. This works in plain regexp:

```js
`(?<QSTR>		// Quoted strings
	(['"\`])
		(?:
			[^\\\2]+?	//  Any character except backslash or the quote character
			|
			\\\\		//  An escaped backslash
			|
			\\[^\2]		//  An escaped character that is not the quote character
			|
			\\(?=\2)	// An escaped quote character, but looking ahead to be able to exit the loop
		)*?				// keep repeating
	\2		//  The closing quote character
)`.replace(/\/\/[^\n]*|\s+/g, '');
```


it works like a charm because the original char from the quote is reused as the \2 so I can ignore other escaped chars (by using it in the `[^\2]`). 

Iny inputs to how I can support this when using the v flag? I can of cause make 3 versions (one for quote char) but that seems like a naive use of the capability of regular expression.

**Comment by slevithan** [comment]
> **Are you saying that with the v flag, I can no longer include a backreference inside a character class?**

No, I'm saying that it is **never** possible to include a backreference inside a character class -- not with `regex`, not with native JavaScript, not with Perl, and not with any flags. Nowhere and never. Alas. Trust me on this for a moment, and I'll get to what your example regex is actually doing in a bit. 😊

Traditionally, backreferences have not been allowed within character classes because character classes can only ever match a single character at a time, whereas a backreference might match any number of characters. But then, JavaScript's `v` introduced the fairly unique concept of allowing character classes to match more than one character at a time (!!!!) via fancy new "properties of strings" referenced via `\p{...}` (as of ES2024 there are only 7 of them and they're all emoji related, with the most useful being `\p{RGI_Emoji}`), as well as via `[\q{...|...}]`. This is advanced stuff, and technically means that the regex engine capabilities are in place to allow backreferences within character classes if the standards/implementation authors wanted to allow that. But as of now, it's not possible.

So what does something like `/()[\1]/` *do*, then? In Unicode-unaware mode, this is an octal escape matching octal character index 1, equivalent to hexadecimal `/()[\x01]/`. With flag `u` or `v`, it is an error, since octal escapes are deprecated. Similarly, `/(?<a>)[\k<a>]/` in Unicode-unaware mode is equivalent to `/(?<a>)[k<a>]/`, which matches a single `k`, `<`, `a`, or `>`. With flag `u` or `v` (which have strict errors for unreserved letter escapes), it is an error, because `\k` is not a known escape within character class context.

> I'm happy it's not the regex library that has a bug, but on the other hand, I felt a bit of excitement when I thought I had found a way to help improve this library.

Haha, I appreciate the sentiment. 😊🙌

> I would love to pick your brain on the original problem, if you dont mind. I would like to match quoted strings ignoring the quote when its being escaped itself. This works in plain regexp: [...]

I'm sorry to say, your example regex does **not** work. It is equivalent to the simplified ``/(['"`])(?:[^\\\x02]|\\\\|\\[^\x02]|\\(?=\1))*?\1/``, which has several issues:

- It incorrectly matches `"\"` and `"\\\"`.
- It does not match quoted strings that include [character index 2](https://codepoints.net/U+0002).
- It includes redundancy. The `\\(?=\1)` alternative will never be used because everything it matches is already matched by `\\[^\x02]`. And the `\\\\` alternative is not needed because `\\[^\x02]` also matches an escaped backslash.
- Unlike my simplified version, it is vulnerable to catastrophic backtracking due to the overlapping matches of the nested `+?` and outer `*?` quantifiers.

But the problem is a good one. How can you use a regex to match a quoted string of type `"..."`, `'...'`, or `` `...` ``, allowing escaped versions of the delimiter within the quoted string, and correctly accounting for escaped backslashes. I have a simple solution for you that I wrote back in 2007. 😊 (You can see I've been at this for a while.) In fact, it's one of my favorite regexes I've come up with, and several others including [Lea Verou](https://www.youtube.com/watch?v=M7vDtxaD7ZU&t=3629s) have called it the most elegant regex they've seen:

``/(["'`])(?:\\?.)*?\1/``

To also avoid an issue with quote marks which are followed by an escaped quote of the same type but are not followed by a closing quote, make the inner quantifier possessive, like so:

`` regex`(?<q>["'\`])(\\?+.)*?\k<q>` ``

Possessive quantifiers aren't supported by native JavaScript, but they are supported by `regex`! Note that, since `regex` implicitly applies flag `n` ("named capture only" mode) by default, I switched to a named capturing group, and used `(...)` for the noncapturing group.

**Comment by slevithan** [comment]
PS: Just because you can't put a backreference in a character class doesn't mean you can't match equivalent strings. For example, instead of `/(...)[a-z\1]+/` you can say `/(...)(?:[a-z]|\1)+/`, and instead of `/(...)[^a-z\1]+/` you can say `/(...)(?:(?!\1)[^a-z])+/`.

**Comment by slevithan** [comment]
PPS: A patch for the issue of being overeager to escape certain characters within character classes (which you helped identify with this issue -- thank you!) was released in v4.1.3. 🙂

**Comment by mathiasrw** [comment]
You make me humble and happy. Thank you!
